demos/sample_azure_iot_pnp/sample_azure_iot_pnp.c

```diff
@@ -0,0 +1,1035 @@
+
+/**
+ * @brief Connect to IoT Hub with reconnection retries.
+ *
+ * If connection fails, retry is attempted after a timeout.
+ * Timeout value will exponentially increase until maximum
+ * timeout value is reached or the number of attempts are exhausted.
+ *
+ * @param pcHostName Hostname of the endpoint to connect to.
+ * @param ulPort Endpoint port.
+ * @param pxNetworkCredentials Pointer to Network credentials.
+ * @param pxNetworkContext Point to Network context created.
+ * @return uint32_t The status of the final connection attempt.
+ */
+static uint32_t prvConnectToServerWithBackoffRetries( const char * pcHostName,
+                                                      uint32_t ulPort,
+                                                      NetworkCredentials_t * pxNetworkCredentials,
+                                                      NetworkContext_t * pxNetworkContext );
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Static buffer used to hold MQTT messages being sent and received.
+ */
+static uint8_t ucMQTTMessageBuffer[ democonfigNETWORK_BUFFER_SIZE ];
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Generate max min payload.
+ */
+int32_t prvInvokeMaxMinCommand( const uint8_t * pucPayload,
+                                uint32_t ulPayloadLength )
+{
+    AzureIoTResult_t xResult;
+    AzureIoTJSONReader_t xReader;
+    uint32_t ulSinceTimeLength;
+    time_t xRawTime;
+    struct tm * pxTimeInfo;
+    size_t xEndTimeLength;
+
+    /* Get the start time */
+    xResult = AzureIoTJSONReader_Init( &xReader, pucPayload, ulPayloadLength );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONReader_NextToken( &xReader );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONReader_TokenIsTextEqual( &xReader, ( uint8_t * ) sampleazureiotCOMMAND_SINCE, strlen( sampleazureiotCOMMAND_SINCE ) );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONReader_NextToken( &xReader );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+    xResult = AzureIoTJSONReader_GetTokenString( &xReader,
+                                                 ucCommandStartTimeValueBuffer,
+                                                 sizeof( ucCommandStartTimeValueBuffer ),
+                                                 &ulSinceTimeLength );
+    configASSERT( xResult == eAzureIoTSuccess );
+
+
+    /* Get the current time as a string. */
+    time( &xRawTime );
+    pxTimeInfo = localtime( &xRawTime );
+    xEndTimeLength = strftime(
+        ucCommandEndTimeValueBuffer,
+        sizeof( ucCommandEndTimeValueBuffer ),
+        sampleazureiotDATE_TIME_FORMAT,
+        pxTimeInfo );
+
+    AzureIoTJSONWriter_t xWriter;
+
+    xResult = AzureIoTJSONWriter_Init( &xWriter, ucCommandPayloadBuffer, sizeof( ucCommandPayloadBuffer ) );
+    configASSERT( xResult == eAzureIoTSuccess );
```

Comment:
its little hard to follow, how this function is communicating result back to caller.  Side effect of this function is some global memory, whose bytes used you are returning.  Better way to express will be to have this function take the reader and output writer.